pyanaconda/ui/webui/__init__.py

```diff
@@ -90,25 +94,59 @@ def run(self):
         """Run the interface."""
         log.debug("web-ui: starting cockpit web view")
 
-        # Force Firefox to be used via the BROWSER environment variable.
-        # This is read by cockpit-desktop and makes it launch Firefox in kiosk mode
-        # instead of the GTK WebKit based web view it launches by default.
-
-        # FIXME: looks like "type" should not be used and _is_live_os is private ?
         if conf.system.provides_liveuser:
-            profile_name = FIREFOX_THEME_LIVE
+            self._watch_webui_on_live()
         else:
-            profile_name = FIREFOX_THEME_DEFAULT
+            self._run_webui()
+
+    def _run_webui(self):
+        # FIXME: This part should be start event loop (could use the WatchProcesses class)
+        # FIXME: We probably want to move this to early execution similar to what we have on live
+        profile_name = FIREFOX_THEME_DEFAULT
 
         proc = startProgram(["/usr/libexec/webui-desktop",
-                            "-t", profile_name, "-r", str(int(self.remote)),
-                            "/cockpit/@localhost/anaconda-webui/index.html"],
+                             "-t", profile_name, "-r", str(int(self.remote)),
+                             "/cockpit/@localhost/anaconda-webui/index.html"],
                             reset_lang=False)
         log.debug("cockpit web view has been started")
-        with open("/run/anaconda/webui_script.pid", "w") as f:
+        with open(self._viewer_pid_file, "w") as f:
             f.write(repr(proc.pid))
         proc.wait()
-        log.debug("cockpit web view has finished running")
+
+    def _watch_webui_on_live(self):
+        """Watch webui-desktop script process on Live.
+
+        It takes long time to start Firefox after the user interaction (on live even 20+ seconds).
+        To avoid that, we are starting the webui-desktop script early in the liveinst script.
+        Here we are just watching if the process is still running. If the browser is closed
+        Anaconda main process will stop.
+        """
+        log.debug("web-ui: watching webui-desktop pid on live environment")
+        pid = -1
+
+        try:
```

Comment:
On a live system, this fails when I test it the following way:
![Screen Shot 2023-10-09 at 11 40 52](https://github.com/rhinstaller/anaconda/assets/14921356/f2802015-141c-4d8c-a05f-2d4e0ed35702)

With these changes I see that this code (watch_webui_on_live) will run before the webui-desktop script will ever rub, so that viewer_pid_file will not be present indeed. When opening a file for reading, python expects the file to be there.

Reply:
This shouldn't be possible, it's sequential order of command execution so the echo command creating a file is executed before Anaconda is even started. The file should be there all the time.

I wonder what is happening.

Reply:
If you take a look on the `liveinst` script you will see the order:

```python
start_anaconda() {
    # early execution of Firefox to reduce starting time
    mkdir -p /run/anaconda
    /usr/libexec/webui-desktop -t live /cockpit/@localhost/anaconda-webui/index.html &
    echo "$!" > /run/anaconda/webui_script.pid

    # start Anaconda main process
    $ANACONDA "$@"
}
```

1. start `webui-desktop` script in background
2. store pid to the pid file
3. start Anaconda backend


I don't think that `anaconda` could be started and initialized before `webui-script.pid` is written.